src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs

```diff
@@ -0,0 +1,314 @@
+        /// <param name="lowerHandle">Lower handle</param>
+        public SNIMarsConnection(SNIHandle lowerHandle)
+        {
+            _lowerHandle = lowerHandle;
+            _lowerHandle.SetAsyncCallbacks(HandleReceiveComplete, HandleSendComplete);
+        }
+
+        public SNIMarsHandle CreateSession(Object callbackObject, bool async)
+        {
+            lock (this)
+            {
+                ushort sessionId = _nextSessionId++;
+                SNIMarsHandle handle = new SNIMarsHandle(this, sessionId, callbackObject, async);
+                _sessions.Add(sessionId, handle);
+                return handle;
+            }
+        }
+
+        /// <summary>
+        /// Start receiving
+        /// </summary>
+        /// <returns></returns>
+        public uint StartReceive()
+        {
+            SNIPacket packet = null;
+
+            if (ReceiveAsync(ref packet) == TdsEnums.SNI_SUCCESS_IO_PENDING)
+            {
+                return TdsEnums.SNI_SUCCESS_IO_PENDING;
+            }
+
+            SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, 0, "Could not start MARS async receive");
+            return TdsEnums.SNI_ERROR;
+        }
+
+        /// <summary>
+        /// Send a packet synchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <returns>SNI error code</returns>
+        public uint Send(SNIPacket packet)
+        {
+            lock (this)
+            {
+                return _lowerHandle.Send(packet);
+            }
+        }
+
+        /// <summary>
+        /// Send a packet asynchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="callback">Completion callback</param>
+        /// <returns>SNI error code</returns>
+        public uint SendAsync(SNIPacket packet, SNIAsyncCallback callback)
+        {
+            lock (this)
+            {
+                return _lowerHandle.SendAsync(packet, callback);
+            }
+        }
+
+        /// <summary>
+        /// Receive a packet asynchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <returns>SNI error code</returns>
+        public uint ReceiveAsync(ref SNIPacket packet)
+        {
+            lock (this)
+            {
+                return _lowerHandle.ReceiveAsync(ref packet);
+            }
+        }
+
+        /// <summary>
+        /// Check SNI handle connection
+        /// </summary>
+        /// <param name="handle"></param>
+        /// <returns>SNI error status</returns>
+        public uint CheckConnection()
+        {
+            lock (this)
+            {
+                return _lowerHandle.CheckConnection();
+            }
+        }
+
+        /// <summary>
+        /// Process a receive error
+        /// </summary>
+        public void HandleReceiveError()
+        {
+            foreach (SNIMarsHandle handle in _sessions.Values)
```

Comment:
_sessions is protected elsewhere by a ```lock(this)```... that's not necessary here?  If it's not necessary because you expect HandleReceiveError will only ever be called while holding the lock, I'd recommend asserting that, e.g. ```Debug.Assert(Monitor.IsEntered(this));```